packages/pyright-internal/src/languageService/completionProvider.ts

```diff
@@ -1427,6 +1456,299 @@ export class CompletionProvider {
+                                return;
+                            }
+                            excludes.add(key);
+
+                            this._addStringLiteralToCompletionList(
+                                key,
+                                quoteValue ? quoteValue.stringValue : undefined,
+                                postText,
+                                quoteValue
+                                    ? quoteValue.quoteCharacter
+                                    : this._parseResults.tokenizerOutput.predominantSingleQuoteCharacter,
+                                completionList
+                            );
+                        }
+                    );
+                });
+            }
+        }
+    }
+
+    private _getTypedDicts(baseExpressionNode: ExpressionNode, parseNode: ParseNode, type: ClassType | UnionType) {
+        if (!(tokenIsAtKey(this) || this._isInDictionaryStringKey(parseNode))) {
+            return [];
+        }
+
+        // path to resolve nested dictionaries
+        let path: DictPathEntry[] = [];
+        let curNode: ParseNode | undefined = parseNode;
+
+        while (curNode && curNode.id !== baseExpressionNode.id) {
+            // List is here as it is a supported container type
+            if (
+                curNode.nodeType === ParseNodeType.Dictionary ||
+                curNode.nodeType === ParseNodeType.Set ||
+                curNode.nodeType === ParseNodeType.List
+            ) {
+                if (
+                    curNode.parent?.nodeType === ParseNodeType.DictionaryKeyEntry &&
+                    curNode.parent?.keyExpression.nodeType === ParseNodeType.StringList
+                ) {
+                    path.push({
+                        entry: curNode.parent,
+                        key: curNode.parent.keyExpression.strings.map((s) => s.value).join(''),
+                    });
+                }
+            }
+
+            curNode = curNode.parent;
+        }
+
+        path = path.reverse();
+        const types = this._getTypedDictsRecursive(baseExpressionNode, parseNode, type, path);
+        const keys = this._getDictExpressionStringKeys(parseNode, new Set([parseNode.parent?.id]));
+        return this._tryNarrowTypedDicts(types, keys);
+
+        function tokenIsAtKey(provider: CompletionProvider): boolean {
+            // this function is only designed return true for states that
+            // are discarded by the parse tree, for example:
+            // {'foo': 'bar', '|}
+            const offset = convertPositionToOffset(provider._position, provider._parseResults.tokenizerOutput.lines)!;
+            const tokens = provider._parseResults.tokenizerOutput.tokens;
+            const index = tokens.getItemAtPosition(offset);
+            const token = tokens.getItemAt(index);
+            const prevToken = tokens.getItemAt(index - 1);
+            return token.type === TokenType.String && prevToken.type === TokenType.Comma;
+        }
```

Comment:
I don't think you need to have `provider` as a parameter here; `this` should work just fine, or just have it be a method like most other helpers.

Reply:
or maybe just pass in `ParseResults` and `position`

Reply:
I got `this` implicitly has type `any` errors.

I was wary about making it a method as this is the part of the code that I am least confident with and thought it might be a misleading helper as it should only be used with surrounding context.

Passing in `ParseResults` and `position` seems like the best solution to me.

Reply:
I see; could probably be a `() =>` function at that point (but Bill's suggestion makes more sense).